mantidproject/mantid
# Adds CheckForMask option + testing for DiffractionFocussing

This adds a `CheckForMask` option to `DiffractionFocussing` that, when set to false, will ensure the number of histograms returned from the `OutputWorkspace` == number of histograms in the grouping scheme.

If an entire group of detectors in a group of the grouping scheme are masked, the current behavior is to not include this focussed histogram in the `OutputWorkspace`. Sometimes, this is undesired. `CheckForMask` allows this to be optional, with the default of `CheckForMask=True` to keep the old behaviour as a default.

The tests for DiffractionFocussing2 in `Framework/Algorithms/test/DiffractionFocussing2Test.h ` have been extended to test for this additional option. Also, the documentation has an extra explanation for the CheckForMask option.

**To test:**

Run the unit tests:
ctest -j2 -R AlgorithmsTest_DiffractionFocussing2

Also, you can run this test script and check output:
```
from __future__ import print_function
from mantid.simpleapi import *

# Create a sample workspace
ws = CreateSampleWorkspace(WorkspaceType="Event", Function="Powder Diffraction", XUnit="dSpacing", NumBanks=10)

# Group Detectors by the 6 Banks
##grouping_ws, spectra_count, group_count = CreateGroupingWorkspace(InstrumentName="NOMAD",GroupDetectorsBy="Group")
grouping_ws, spectra_count, group_count = CreateGroupingWorkspace(InputWorkspace=ws,GroupDetectorsBy="bank")

# Create map of { 'Group #' : 'List of detector ids' }
group2udets = dict()
for i, gid_array in enumerate(grouping_ws.extractY()):
    gid = gid_array[0]
    if gid not in group2udets:
        group2udets[gid] = list()                                                                     
    group2udets[gid].append(i)

print("Number of groups: %d" % group_count)

# Focus using the grouping
focussed_ws = DiffractionFocussing(InputWorkspace=ws, GroupingWorkspace=grouping_ws)
print("Number of spectra (Focussed): %d" % focussed_ws.getNumberHistograms())

# Mask an entire Group
mask_group_id = 1
mask_wksp_id_list = group2udets[mask_group_id]
MaskDetectors(Workspace=ws, WorkspaceIndexList=mask_wksp_id_list)

# Focus using the mask + grouping
masked_focussed_ws = DiffractionFocussing(InputWorkspace=ws,
                                          GroupingWorkspace=grouping_ws)
print("Number of spectra (Masked + Focussed): %d" % masked_focussed_ws.getNumberHistograms())

# Focus using the mask + grouping but with CheckForMask == False
masked_focussed_ws2 = DiffractionFocussing(InputWorkspace=ws,
                                           GroupingWorkspace=grouping_ws,
                                           CheckForMask=False)
print("Number of spectra (Masked + Focussed w/o Mask Check): %d" % masked_focussed_ws2.getNumberHistograms())
```

Output:
```
Number of groups: 10
Number of spectra (Focussed): 10
Number of spectra (Masked + Focussed): 9
Number of spectra (Masked + Focussed w/o Mask Check): 10
```

Fixes #22602 

Does this update require release notes?

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
